Logically identical code. One compiles, one does not

I tried to compile this code [playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3a7b73546d419d316ad792d0c6362152):

```
use std::time::{Duration, Instant};
use std::thread;

fn time_bounded_retry<T, E>(
    max_time: Duration,
    sleep_for: Duration,
    f: impl Fn() -> Result<T, E>,
) -> Result<T, E> {
    let start = Instant::now();
    while let result = f() {
        match result {
            Ok(_) => return result,
            Err(_) => {
                if start.elapsed() > max_time {
                    return result;
                }

                thread::sleep(sleep_for);
            }
        }
    }
    // unreachable!()
}
```

However, it produces an error. If you uncomment the `unreachable!()` will work. Here's the error message:

```
    |
191 |     while let result = f() {
    |     ^^^^^^^^^^^^^^^^^^^^^^ this might have zero elements to iterate on
192 |         match result {
193 |             Ok(_) => return result,
    |                      ------------- if the loop doesn't execute, this value would never get returned
...
196 |                     return result;
    |                     ------------- if the loop doesn't execute, this value would never get returned
    = help: return a value for the case when the loop has zero elements to iterate on, or consider changing the return type to account for that possibility
help: try adding an expression at the end of the block
```

This code is functionally identical to the `while let` code and can compile without an `unreachable!()`:

```
fn time_bounded_retry<T, E>(
    max_time: Duration,
    sleep_for: Duration,
    f: impl Fn() -> Result<T, E>,
) -> Result<T, E> {
    let start = Instant::now();
    loop {
        let result = f();
        match result {
            Ok(_) => return result,
            Err(_) => {
                if start.elapsed() > max_time {
                    return result;
                }

                thread::sleep(sleep_for);
            }
        }
    }
}
```

The code in the while loop is irrefutable, so it is functionally the same as the `loop` case. Even if it's not a good practice to use an irrefutable statement in a while loop (and clippy will catch this) I think it should still compile.

My main reason for wanting it to compile is that while iterating, I wrote the `while let` version first, and because I could not get it to compile without the `unreachable!()` statement mistakenly believed that the compiler would not allow something like the second result. As a result of that mistaken impression ended up with this code:

```
fn time_bounded_retry<T, E>(
    max_time: Duration,
    sleep_for: Duration,
    f: impl Fn() -> Result<T, E>,
) -> Result<T, E> {
    let start = Instant::now();
    let mut result;
    loop {
        result = f();
        if result.is_ok() || start.elapsed() > max_time {
            break;
        }
        thread::sleep(sleep_for);
    }
    result
}
```

Which is not ideal (needs a temp mutable variable). A co-worker without the above learned helplessness `unreachable!()` experience wrote this code below, which I believe is a much better way to implement the logic as it does not require a temp mutable variable:


```
fn time_bounded_retry<T, E>(
    max_time: Duration,
    sleep_for: Duration,
    f: impl Fn() -> Result<T, E>,
) -> Result<T, E> {
    let start = Instant::now();
    loop {
        result = f();
        if result.is_ok() || start.elapsed() > max_time {
            return result;
        }
        thread::sleep(sleep_for);
    }
}
```

So effectively, I'm advocating for `while let var = //...` to be treated the same as `loop { let var = //...`, not because I think it's a good idea, but because having one error and not the other gives false feedback to the coder.

### Meta

<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
$ rustc --version --verbose
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-apple-darwin
release: 1.73.0
LLVM version: 17.0.2
```

Quick triage note before committing — on a stage1 build of current `main` (`rustc 1.97.0-dev`), @traviscross's minimization now produces:

```text
error[E0308]: mismatched types
 --> src/lib.rs:2:5
  |
1 |   fn foo() -> bool {
  |               ---- expected `bool` because of return type
2 | /     while let x = false {
3 | |         if x { return true }
4 | |     }
  | |_____^ expected `bool`, found `()`
  |
  = note: `while` loops evaluate to unit type `()`
help: consider returning a value here
  |
4 ~     }
5 +     /* `bool` value */
  |
```

The misleading "this might have zero elements to iterate on" / "if the loop doesn't execute" framing from the original 2023 report is gone — the diagnostic now correctly explains that `while` loops evaluate to `()` and presents a normal return-type mismatch.

What still seems off: `while let x = false` is irrefutable, so the loop is effectively `loop { let x = …; … }`, the function body is diverging, and the suggestion to add a tail expression is misleading (there's no execution path to it). But addressing that appears to require detecting irrefutable `while let` patterns during type-checking — which feels more substantial than typical E-easy work to me.

Two quick questions before I dig further:
1. Should the `E-easy` label be reconsidered given the current state?
2. Or has the original concern been addressed enough to close this in favor of a narrower follow-up?

Either answer is useful — just want to make sure I'm scoped correctly before claiming.

--- a/compiler/rustc_hir_typeck/src/coercion.rs
+++ b/compiler/rustc_hir_typeck/src/coercion.rs
@@ -1895,11 +1895,11 @@ impl<'tcx> CoerceMany<'tcx> {
 
                         err.note(format!("{loop_type} evaluate to unit type `()`"));
                         if loop_src == hir::LoopSource::While
-                            && let Some(pat) = irrefutable_while_let_pattern(block)
+                            && let Some(pat) = irrefutable_if_let_expr(block)
                         {
                             err.span_note(
                                 pat.span,
-                                "this pattern always matches, so the loop condition never fails",
+                                "this pattern always matches, consider using `loop` instead",
                             );
                         }
                     }
@@ -2134,7 +2134,7 @@ impl<'tcx> CoerceMany<'tcx> {
     }
 }
 
-fn irrefutable_while_let_pattern<'hir>(block: &hir::Block<'hir>) -> Option<&'hir hir::Pat<'hir>> {
+fn irrefutable_if_let_expr<'hir>(block: &hir::Block<'hir>) -> Option<&'hir hir::Pat<'hir>> {
     let hir::ExprKind::If(cond, _, _) = block.expr?.kind else {
         return None;
     };

--- a/compiler/rustc_hir_typeck/src/coercion.rs
+++ b/compiler/rustc_hir_typeck/src/coercion.rs
@@ -1897,7 +1897,7 @@ impl<'tcx> CoerceMany<'tcx> {
                         if loop_src == hir::LoopSource::While
                             && let Some(pat) = irrefutable_if_let_expr(block)
                         {
-                            err.span_note(
+                            err.span_label(
                                 pat.span,
                                 "this pattern always matches, consider using `loop` instead",
                             );

--- a/compiler/rustc_hir_typeck/src/coercion.rs
+++ b/compiler/rustc_hir_typeck/src/coercion.rs
@@ -1881,7 +1881,7 @@ impl<'tcx> CoerceMany<'tcx> {
 
                 if let Some(expr) = expression {
                     if let hir::ExprKind::Loop(
-                        _,
+                        block,
                         _,
                         loop_src @ (hir::LoopSource::While | hir::LoopSource::ForLoop),
                         _,
@@ -1894,6 +1894,14 @@ impl<'tcx> CoerceMany<'tcx> {
                         };
 
                         err.note(format!("{loop_type} evaluate to unit type `()`"));
+                        if loop_src == hir::LoopSource::While
+                            && let Some(pat) = irrefutable_while_let_pattern(block)
+                        {
+                            err.span_note(
+                                pat.span,
+                                "this pattern always matches, so the loop condition never fails",
+                            );
+                        }
                     }
 
                     fcx.emit_coerce_suggestions(
@@ -2123,6 +2131,24 @@ impl<'tcx> CoerceMany<'tcx> {
             assert!(self.expressions.is_empty());
             fcx.tcx.types.never
         }
+    }
+}
+
+fn irrefutable_while_let_pattern<'hir>(block: &hir::Block<'hir>) -> Option<&'hir hir::Pat<'hir>> {
+    let hir::ExprKind::If(cond, _, _) = block.expr?.kind else {
+        return None;
+    };
+    let hir::ExprKind::Let(let_expr) = cond.kind else {
+        return None;
+    };
+    simple_irrefutable_pattern(let_expr.pat).then_some(let_expr.pat)
+}
+
+fn simple_irrefutable_pattern(pat: &hir::Pat<'_>) -> bool {
+    match pat.kind {
+        hir::PatKind::Wild | hir::PatKind::Binding(_, _, _, None) => true,
+        hir::PatKind::Tuple(pats, _) => pats.iter().all(simple_irrefutable_pattern),
+        _ => false,
     }
 }
 

--- a/tests/ui/coercion/coerce-loop-issue-122561.rs
+++ b/tests/ui/coercion/coerce-loop-issue-122561.rs
@@ -69,6 +69,24 @@ fn while_zero_times() -> bool {
     }
 }
 
+fn while_let_binding() -> bool {
+    while let x = false {
+    //~^ ERROR mismatched types
+        if x {
+            return true;
+        }
+    }
+}
+
+fn while_let_tuple() -> bool {
+    while let (x, _) = (false, true) {
+    //~^ ERROR mismatched types
+        if x {
+            return true;
+        }
+    }
+}
+
 fn while_never_type() -> ! {
     while true {
     //~^ ERROR mismatched types

--- a/tests/ui/coercion/coerce-loop-issue-122561.stderr
+++ b/tests/ui/coercion/coerce-loop-issue-122561.stderr
@@ -173,7 +173,10 @@ error[E0308]: mismatched types
    |
 LL |   fn while_let_binding() -> bool {
    |                             ---- expected `bool` because of return type
-LL | /     while let x = false {
+LL |       while let x = false {
+   |       ^         - this pattern always matches, consider using `loop` instead
+   |  _____|
+   | |
 LL | |
 LL | |         if x {
 LL | |             return true;
@@ -182,11 +185,6 @@ LL | |     }
    | |_____^ expected `bool`, found `()`
    |
    = note: `while` loops evaluate to unit type `()`
-note: this pattern always matches, consider using `loop` instead
-  --> $DIR/coerce-loop-issue-122561.rs:73:15
-   |
-LL |     while let x = false {
-   |               ^
 help: consider returning a value here
    |
 LL ~     }
@@ -198,7 +196,10 @@ error[E0308]: mismatched types
    |
 LL |   fn while_let_tuple() -> bool {
    |                           ---- expected `bool` because of return type
-LL | /     while let (x, _) = (false, true) {
+LL |       while let (x, _) = (false, true) {
+   |       ^         ------ this pattern always matches, consider using `loop` instead
+   |  _____|
+   | |
 LL | |
 LL | |         if x {
 LL | |             return true;
@@ -207,11 +208,6 @@ LL | |     }
    | |_____^ expected `bool`, found `()`
    |
    = note: `while` loops evaluate to unit type `()`
-note: this pattern always matches, consider using `loop` instead
-  --> $DIR/coerce-loop-issue-122561.rs:82:15
-   |
-LL |     while let (x, _) = (false, true) {
-   |               ^^^^^^
 help: consider returning a value here
    |
 LL ~     }

--- a/tests/ui/coercion/coerce-loop-issue-122561.stderr
+++ b/tests/ui/coercion/coerce-loop-issue-122561.stderr
@@ -182,7 +182,7 @@ LL | |     }
    | |_____^ expected `bool`, found `()`
    |
    = note: `while` loops evaluate to unit type `()`
-note: this pattern always matches, so the loop condition never fails
+note: this pattern always matches, consider using `loop` instead
   --> $DIR/coerce-loop-issue-122561.rs:73:15
    |
 LL |     while let x = false {
@@ -207,7 +207,7 @@ LL | |     }
    | |_____^ expected `bool`, found `()`
    |
    = note: `while` loops evaluate to unit type `()`
-note: this pattern always matches, so the loop condition never fails
+note: this pattern always matches, consider using `loop` instead
   --> $DIR/coerce-loop-issue-122561.rs:82:15
    |
 LL |     while let (x, _) = (false, true) {

--- a/tests/ui/coercion/coerce-loop-issue-122561.stderr
+++ b/tests/ui/coercion/coerce-loop-issue-122561.stderr
@@ -7,7 +7,7 @@ LL |     while true {
    = note: `#[warn(while_true)]` on by default
 
 warning: denote infinite loops with `loop { ... }`
-  --> $DIR/coerce-loop-issue-122561.rs:73:5
+  --> $DIR/coerce-loop-issue-122561.rs:91:5
    |
 LL |     while true {
    |     ^^^^^^^^^^ help: use `loop`
@@ -171,6 +171,56 @@ LL +     /* `bool` value */
 error[E0308]: mismatched types
   --> $DIR/coerce-loop-issue-122561.rs:73:5
    |
+LL |   fn while_let_binding() -> bool {
+   |                             ---- expected `bool` because of return type
+LL | /     while let x = false {
+LL | |
+LL | |         if x {
+LL | |             return true;
+LL | |         }
+LL | |     }
+   | |_____^ expected `bool`, found `()`
+   |
+   = note: `while` loops evaluate to unit type `()`
+note: this pattern always matches, so the loop condition never fails
+  --> $DIR/coerce-loop-issue-122561.rs:73:15
+   |
+LL |     while let x = false {
+   |               ^
+help: consider returning a value here
+   |
+LL ~     }
+LL +     /* `bool` value */
+   |
+
+error[E0308]: mismatched types
+  --> $DIR/coerce-loop-issue-122561.rs:82:5
+   |
+LL |   fn while_let_tuple() -> bool {
+   |                           ---- expected `bool` because of return type
+LL | /     while let (x, _) = (false, true) {
+LL | |
+LL | |         if x {
+LL | |             return true;
+LL | |         }
+LL | |     }
+   | |_____^ expected `bool`, found `()`
+   |
+   = note: `while` loops evaluate to unit type `()`
+note: this pattern always matches, so the loop condition never fails
+  --> $DIR/coerce-loop-issue-122561.rs:82:15
+   |
+LL |     while let (x, _) = (false, true) {
+   |               ^^^^^^
+help: consider returning a value here
+   |
+LL ~     }
+LL +     /* `bool` value */
+   |
+
+error[E0308]: mismatched types
+  --> $DIR/coerce-loop-issue-122561.rs:91:5
+   |
 LL |   fn while_never_type() -> ! {
    |                            - expected `!` because of return type
 LL | /     while true {
@@ -188,7 +238,7 @@ LL +     /* `loop {}` or `panic!("...")` */
    |
 
 error[E0308]: mismatched types
-  --> $DIR/coerce-loop-issue-122561.rs:87:5
+  --> $DIR/coerce-loop-issue-122561.rs:105:5
    |
 LL | /     for i in 0.. {
 LL | |
@@ -203,7 +253,7 @@ LL +     /* `i32` value */
    |
 
 error[E0308]: mismatched types
-  --> $DIR/coerce-loop-issue-122561.rs:94:9
+  --> $DIR/coerce-loop-issue-122561.rs:112:9
    |
 LL | /         for i in 0..5 {
 LL | |
@@ -218,7 +268,7 @@ LL +         /* `usize` value */
    |
 
 error[E0308]: mismatched types
-  --> $DIR/coerce-loop-issue-122561.rs:100:9
+  --> $DIR/coerce-loop-issue-122561.rs:118:9
    |
 LL | /         while false {
 LL | |
@@ -233,7 +283,7 @@ LL +         /* `usize` value */
    |
 
 error[E0308]: mismatched types
-  --> $DIR/coerce-loop-issue-122561.rs:106:23
+  --> $DIR/coerce-loop-issue-122561.rs:124:23
    |
 LL |     let _ = |a: &[(); for x in 0..2 {}]| {};
    |                       ^^^^^^^^^^^^^^^^ expected `usize`, found `()`
@@ -244,6 +294,6 @@ help: consider returning a value here
 LL |     let _ = |a: &[(); for x in 0..2 {} /* `usize` value */]| {};
    |                                        +++++++++++++++++++
 
-error: aborting due to 14 previous errors; 2 warnings emitted
+error: aborting due to 16 previous errors; 2 warnings emitted
 
 For more information about this error, try `rustc --explain E0308`.

